### PR TITLE
Add std::hash overloads for Node, TNode and TypeNode.

### PR DESCRIPTION
**Pull request by mpreiner**
Eliminates NodeHashFunction, TNodeHashFunction and TypeNodeHashFunction.

Note to reviewers: Main changes are in `expr/node.{h,cpp}` and `expr/type_node.{h,cpp}`.

### CODE REVIEW

File: src/theory/substitutions.h [hunk 1]
```diff
@@ -43,33 +43,33 @@ namespace theory {
...
 
-  typedef context::CDHashMap<Node, Node, NodeHashFunction> NodeMap;
-
-  typedef NodeMap::iterator iterator;
-  typedef NodeMap::const_iterator const_iterator;
+ typedef NodeMap::iterator iterator;
+ typedef NodeMap::const_iterator const_iterator;
 
 private:
+ typedef std::unordered_map<Node, Node> NodeCache;
```
Comment (4tXJ7f) [inline src/theory/substitutions.h:52]: Looks like the formatting is wrong because of `private`/`public` not being indented.

**Review by 4tXJ7f — APPROVED** [review]
LGTM, just one minor formatting issue.

File: src/theory/substitutions.h (hunk 1 above)
Comment (mpreiner) [reply src/theory/substitutions.h]: clang-formatted the whole class, should be fixed now.

**Review by nafur — APPROVED** [review]
TBH I've only looked at  node/type_node. But that LGTM.

**Pull request merged**